### ISSUE TITLE
Requests can hang with malformed JSONs; depends on error handling

### ISSUE THREAD

**Issue by juona** [jaydenseric/graphql-upload#123]
Jayden,

Great work on the library, and thanks for pushing guys @ Apollo. Your workaround works well enough so that I wouldn't care about them fixing their own dependency.

I am now using graphql-upload@8.0.1 with Express and have found this one rather small issue.

In the express middleware you have written code which (please correct me if I am wrong) is meant to delay the final response until the upload has ended:

- https://github.com/jaydenseric/graphql-upload/blob/master/src/graphqlUploadExpress.mjs#L36

- https://github.com/jaydenseric/graphql-upload/blob/master/src/graphqlUploadExpress.mjs#L32.

The issue is that the `finished` promise is listening for `request.on("end")`, which apparently doesn't happen if an error occurs while uploading. Scenario is this:

1. Send a malformed request (e.q. an invalid `operations` JSON).
2. Have an error handler, which uses `request.send` (or `.json` - it resolves to `.send`). E.g. I have this little piece:

```
app.use((err, _, res, __) => {
  res.status(err.status).json({
    message: httpStatus[err.status],
    stack: config.printErrorStack ? err.stack : {}
  });
});
```

When `.json()` is called, the flow goes to your middleware:

https://github.com/jaydenseric/graphql-upload/blob/master/src/graphqlUploadExpress.mjs#L36

However, in this scenario the `end` event is never fired on the `request`. Two workarounds are available:

1. Use `response.end()` instead of `.send()` or `.json()`. I'd like to stick with `.json`, however.
2. Manually emit the `end` event in case of an error: `request.emit("end")`.

Questions to you:

1. Do I understand your code correctly?
2. Do you know why the `end` event is not firing in case of an error in the upload middleware?
3. Does emitting `end` manually seem like a good fix which can be applied now?
4. Would you fix this? I guess you could do this using processRequest.catch() but I am not sure what your desired outcome would be in this case.

**Comment by mike-marcacci** [comment]
1. Of course!
2. No worries at all – Jayden is certainly the driving force behind this project, and the one who gets stuck with things when the rest of us bail ;-) so you were totally correct to address him.
3. Thanks!
4. I am seeing this as well. Changing [this line](https://github.com/jaydenseric/graphql-upload/blob/e0ff3d09018ef9d725850643c278a51a9ae03765/src/test.mjs#L156) to:
    ```js
    body.append('1', 'a'.repeat(70000), { filename: 'a.txt' })
    ```
    makes the payload is larger than a single TCP packet and larger than the node stream's internal buffer, and ensures that the HTTP request isn't fully consumed when the error occurs. Under these conditions, I can see exactly what you are describing. I'm considering having all the tests run with a small and large payload, to help us catch these issues in advance.
    
    I can see that the stream is *not* paused when `processRequest` calls `.resume` but is paused again when the error middleware is called. This appears to share the same behavior with koa, actually, so the bug is either on our end, or there's something unexpected going on with node's http library. I'm going to dig into this right now and post back.

What I mean by "broken HTTP transaction, is that the client will attempt to send its headers and the entirety of its request body along the TCP connection (this is all assuming we aren't using the HTTP 100 [Continue] flow). If we attempt to send a response and end the request/close the connection before the body has finished being sent, all browsers and all popular reverse proxies will treat this as a network failure, and the response status and body will be abandoned. While this is fine if you're trying to prevent DOS attacks, you definitely don't want this kind of behavior for a malicious client. There is the case where node has received the entire request body (so the client will now accept a response) even though our code hasn't actually read the entire body out of node's internal stream buffer. This is essentially why this bug exists – because I didn't test this case with a large enough payload. :(

**Comment by juona** [comment]
Thank you, the explanation makes sense : ]

Come to think of it, I did notice that smaller JPEGs seemed to work differently, i.e. would sometimes return responses, but the behaviour was inconsistent and I couldn't understand what was going on, so I didn't mention it in the description.

Happy to see smart people working on packages that I use - makes me feel secure ^^

**Comment by jaydenseric** [comment]
The fix is published in [`graphql-upload@8.0.2`](https://github.com/jaydenseric/graphql-upload/releases/tag/v8.0.2), thanks all 🙌
